### PR TITLE
lib/backup/azremote: fix copying for parts larger than 256M by using async copy

### LINKED ISSUE #3477: vmbackup error when specifying -origin and using Azure Storage Account
**Describe the bug**
When using the vmbackup tool with an Azure Storage account as a storage medium, and specifying the `-origin` flag to point to an existing (incremental) backup to speed the process up via a server-side copy, the tool encounters a 409 error from Azure. (Logs below). 

It is unclear if this is a result of the way that the vmbackup tool interacts with the Azure Storage account, or if it could be a permissions/quota issue on my Azure account. 

**To Reproduce**
I have two K8s cron jobs setup, one performs incremental backups (by always pointing to the same folder in the storage account) daily, then there is a weekly job that points to a different path, and speicfies the `-origin` flag pointing to the incremental backup, at `/latest`. Example `vmbackup` command:

```
the_date=$(date +%F)
./vmbackup-prod \
-storageDataPath=/storage \
-snapshot.createURL=http://victoria-metrics-victoria-metrics-cluster-vmstorage-1.victoria-metrics-victoria-metrics-cluster-vmstorage.victoria-metrics:8482/snapshot/create \
-dst=azblob://victoria-metrics/full/${the_date}/victoria-metrics-victoria-metrics-cluster-vmstorage-1 \
-origin=azblob://victoria-metrics/latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1
```

Credentials are provided via the `AZURE_STORAGE_ACCOUNT_CONNECTION_STRING` env var. 

The backup at the `-origin` location is about 500GB.

**Expected behavior**
The backup should succeed using a server-side copy when `-origin` is specified. 

**Logs**
```
➜  ~ k8s:(mr-admin-A) k logs -n victoria-metrics vmstorage-1-full-manual-ldjmk -f
2022-12-12T19:43:51.912Z	info	VictoriaMetrics/lib/logger/flag.go:12	build version: vmbackup-20221007-005245-tags-v1.82.0-0-g832276064
2022-12-12T19:43:51.912Z	info	VictoriaMetrics/lib/logger/flag.go:13	command-line flags
2022-12-12T19:43:51.912Z	info	VictoriaMetrics/lib/logger/flag.go:20	  -dst="azblob://victoria-metrics/full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1"
2022-12-12T19:43:51.912Z	info	VictoriaMetrics/lib/logger/flag.go:20	  -origin="azblob://victoria-metrics/latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1"
2022-12-12T19:43:51.912Z	info	VictoriaMetrics/lib/logger/flag.go:20	  -snapshot.createURL="http://victoria-metrics-victoria-metrics-cluster-vmstorage-1.victoria-metrics-victoria-metrics-cluster-vmstorage.victoria-metrics:8482/snapshot/create"
2022-12-12T19:43:51.912Z	info	VictoriaMetrics/lib/logger/flag.go:20	  -storageDataPath="/storage"
2022-12-12T19:43:51.912Z	info	VictoriaMetrics/app/vmbackup/main.go:53	Snapshot create url http://victoria-metrics-victoria-metrics-cluster-vmstorage-1.victoria-metrics-victoria-metrics-cluster-vmstorage.victoria-metrics:8482/snapshot/create
2022-12-12T19:43:51.912Z	info	VictoriaMetrics/app/vmbackup/main.go:60	Snapshot delete url http://victoria-metrics-victoria-metrics-cluster-vmstorage-1.victoria-metrics-victoria-metrics-cluster-vmstorage.victoria-metrics:8482/snapshot/delete
2022-12-12T19:43:51.912Z	info	VictoriaMetrics/lib/snapshot/snapshot.go:28	Creating snapshot
2022-12-12T19:43:53.582Z	info	VictoriaMetrics/lib/snapshot/snapshot.go:52	Snapshot 20221212194351-172EA3346AA94EFA created
2022-12-12T19:43:53.583Z	info	VictoriaMetrics/lib/httpserver/httpserver.go:96	starting http server at http://127.0.0.1:8420/
2022-12-12T19:43:53.583Z	info	VictoriaMetrics/lib/httpserver/httpserver.go:97	pprof handlers are exposed at http://127.0.0.1:8420/debug/pprof/
2022-12-12T19:43:53.623Z	error	VictoriaMetrics/lib/backup/azremote/azblob.go:293	GetProperties("https://mradminmetricsbackups.blob.core.windows.net/victoria-metrics/full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/backup_complete.ignore") returned HEAD https://mradminmetricsbackups.blob.core.windows.net/victoria-metrics/full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/backup_complete.ignore
--------------------------------------------------------------------------------
RESPONSE 404: 404 The specified blob does not exist.
ERROR CODE: BlobNotFound
--------------------------------------------------------------------------------
Response contained no body
--------------------------------------------------------------------------------
2022-12-12T19:43:53.623Z	info	VictoriaMetrics/lib/backup/actions/backup.go:72	starting backup from fslocal "/storage/snapshots/20221212194351-172EA3346AA94EFA" to AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} using origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.633Z	info	VictoriaMetrics/lib/backup/actions/backup.go:78	obtained 1447 parts from src fslocal "/storage/snapshots/20221212194351-172EA3346AA94EFA"
2022-12-12T19:43:53.665Z	info	VictoriaMetrics/lib/backup/actions/backup.go:84	obtained 136 parts from dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.935Z	info	VictoriaMetrics/lib/backup/actions/backup.go:90	obtained 1494 parts from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.940Z	info	VictoriaMetrics/lib/backup/actions/backup.go:122	server-side copying 1136 parts from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.941Z	info	VictoriaMetrics/lib/backup/actions/backup.go:125	server-side copying part{path: "data/small/2022_09/535632568_4624033_20220930015512.441_20220930072925.800_17171BEA2F769ED7/values.bin", file_size: 114870012, offset: 0, size: 114870012} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.941Z	info	VictoriaMetrics/lib/backup/actions/backup.go:125	server-side copying part{path: "data/small/2022_09/19022_19022_20220915205131.856_20220930235958.277_17171BEA2F76CE7B/timestamps.bin", file_size: 0, offset: 0, size: 0} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.941Z	info	VictoriaMetrics/lib/backup/actions/backup.go:125	server-side copying part{path: "data/small/2022_09/19022_19022_20220915205131.856_20220930235958.277_17171BEA2F76CE7B/metaindex.bin", file_size: 731, offset: 0, size: 731} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.941Z	info	VictoriaMetrics/lib/backup/actions/backup.go:125	server-side copying part{path: "data/small/2022_09/535632568_4624033_20220930015512.441_20220930072925.800_17171BEA2F769ED7/timestamps.bin", file_size: 428092536, offset: 0, size: 428092536} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.941Z	info	VictoriaMetrics/lib/backup/actions/backup.go:125	server-side copying part{path: "data/small/2022_09/535632568_4624033_20220930015512.441_20220930072925.800_17171BEA2F769ED7/index.bin", file_size: 42583330, offset: 0, size: 42583330} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.941Z	info	VictoriaMetrics/lib/backup/actions/backup.go:125	server-side copying part{path: "data/small/2022_09/19022_19022_20220915205131.856_20220930235958.277_17171BEA2F76CE7B/index.bin", file_size: 125788, offset: 0, size: 125788} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.941Z	info	VictoriaMetrics/lib/backup/actions/backup.go:125	server-side copying part{path: "data/small/2022_09/19022_19022_20220915205131.856_20220930235958.277_17171BEA2F76CE7B/values.bin", file_size: 0, offset: 0, size: 0} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.942Z	info	VictoriaMetrics/lib/backup/actions/backup.go:125	server-side copying part{path: "data/small/2022_09/504495497_4621409_20220930105117.320_20220930134724.755_17171BEA2F76B109/timestamps.bin", file_size: 362707169, offset: 0, size: 362707169} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.942Z	info	VictoriaMetrics/lib/backup/actions/backup.go:125	server-side copying part{path: "data/small/2022_09/31_10_20220930235749.326_20220930235946.718_17171BEA2F76CE9B/metaindex.bin", file_size: 66, offset: 0, size: 66} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.942Z	info	VictoriaMetrics/lib/backup/actions/backup.go:125	server-side copying part{path: "data/small/2022_09/549208699_4614651_20220930173916.269_20220930214305.876_17171BEA2F76C7E1/timestamps.bin", file_size: 397423708, offset: 0, size: 397423708} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:53.986Z	info	VictoriaMetrics/lib/backup/actions/backup.go:125	server-side copying part{path: "data/small/2022_09/31_10_20220930235749.326_20220930235946.718_17171BEA2F76CE9B/timestamps.bin", file_size: 0, offset: 0, size: 0} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}
2022-12-12T19:43:59.142Z	info	VictoriaMetrics/lib/backup/actions/backup.go:133	server-side copied 8 out of 1136 parts from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} in 5.201486629s
2022-12-12T19:43:59.143Z	fatal	VictoriaMetrics/app/vmbackup/main.go:105	cannot create backup: cannot copy part{path: "data/small/2022_09/504495497_4621409_20220930105117.320_20220930134724.755_17171BEA2F76B109/timestamps.bin", file_size: 362707169, offset: 0, size: 362707169} from origin AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to dst AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}: cannot copy "data/small/2022_09/504495497_4621409_20220930105117.320_20220930134724.755_17171BEA2F76B109/timestamps.bin" from AZBlob{container: "victoria-metrics", dir: "latest/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"} to AZBlob{container: "victoria-metrics", dir: "full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/"}: PUT https://mradminmetricsbackups.blob.core.windows.net/victoria-metrics/full/2022-12-12/victoria-metrics-victoria-metrics-cluster-vmstorage-1/data/small/2022_09/504495497_4621409_20220930105117.320_20220930134724.755_17171BEA2F76B109/timestamps.bin/00000000159E78E1_0000000000000000_00000000159E78E1
--------------------------------------------------------------------------------
RESPONSE 409: 409 The source request body for synchronous copy is too large and exceeds the maximum permissible limit (256MB).
ERROR CODE: CannotVerifyCopySource
--------------------------------------------------------------------------------
<?xml version="1.0" encoding="utf-8"?><Error><Code>CannotVerifyCopySource</Code><Message>The source request body for synchronous copy is too large and exceeds the maximum permissible limit (256MB).
RequestId:c6e898b5-e01e-000e-7f62-0e7906000000
Time:2022-12-12T19:43:53.9944153Z</Message></Error>
```

**Screenshots**
n/a

**Version**
I am using this image: `victoriametrics/vmbackup:v1.82.0`

**Used command-line flags**
Command can be seen above.

### PR DESCRIPTION
**Pull request by zekker6**
Azure API limits copying of files larger than 256M in a sync mode, this PR implements using async mode for parts copy.

Fixes  #3477

### CODE REVIEW

File: lib/backup/azremote/azblob.go [hunk 1]
```diff
@@ -178,12 +178,33 @@ func (fs *FS) CopyPart(srcFS common.OriginFS, p common.Part) error {
...
+	for {
+		r, err := dbc.GetProperties(ctx, nil)
+		if err != nil {
+			return fmt.Errorf("failed to check copy status, cannot get properties of %q at %s: %w", p.Path, fs, err)
+		}
+
+		// After the copy will be finished status will be changed to success/failed/aborted
+		// Ref: https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob-properties#response-headers - x-ms-copy-status
+		if *r.CopyStatus != blob.CopyStatusTypePending {
+			copyStatus = r.CopyStatus
```
Comment (hagen1778) [inline lib/backup/azremote/azblob.go:200]: should we also store `x-ms-copy-status-description` for debugging purpose?
```
Appears only when x-ms-copy-status is failed or pending. Describes the cause of a fatal or non-fatal copy operation failure.
```

**Review by hagen1778 — APPROVED** [review]

File: lib/backup/azremote/azblob.go (hunk 1 above)
Comment (zekker6) [reply lib/backup/azremote/azblob.go:201]: Thanks for suggestion, implemented this.

File: lib/backup/azremote/azblob.go [hunk 2]
```diff
@@ -178,12 +178,35 @@ func (fs *FS) CopyPart(srcFS common.OriginFS, p common.Part) error {
...
+			copyStatusDescription = r.CopyStatusDescription
+			break
+		}
+		time.Sleep(5 * time.Second)
```
Comment (f41gh7) [inline lib/backup/azremote/azblob.go:205]: Probably, it's possible to remove sleep and use some polling/watching for changes?

File: lib/backup/azremote/azblob.go [hunk 3]
```diff
@@ -178,12 +178,35 @@ func (fs *FS) CopyPart(srcFS common.OriginFS, p common.Part) error {
...
 	}
 
-	// Hotfix for SDK issue: https://github.com/Azure/azure-sdk-for-go/issues/19245
-	t = strings.Replace(t, "/?", "?", -1)
 	ctx := context.Background()
-	_, err = dbc.CopyFromURL(ctx, t, &blob.CopyFromURLOptions{})
+
+	// In order to support copy of files larger than 256MB, we need to use the async copy
+	// Ref: https://learn.microsoft.com/en-us/rest/api/storageservices/copy-blob-from-url
+	_, err = dbc.StartCopyFromURL(ctx, t, &blob.StartCopyFromURLOptions{})
```
Comment (f41gh7) [inline lib/backup/azremote/azblob.go:185]: should we use it only for parts > 256MB? and use sync method for parts up to 256mb?

File: lib/backup/azremote/azblob.go (hunk 2 above)
Comment (zekker6) [reply lib/backup/azremote/azblob.go:205]: Unfortunately API has no method which implements that.

File: lib/backup/azremote/azblob.go (hunk 3 above)
Comment (zekker6) [reply lib/backup/azremote/azblob.go:185]: It seems to me like using single flow for all files should better in terms of code complexity and will be easier to maintain.
Implementing 2 methods might require us updating code when/if azure limits for blocking method will change, while always using async copy won't require any additional changes.

**Review by valyala — APPROVED** [review]

**Comment by valyala** [comment]
@zekker6 , thanks for the bugfix!

**Pull request merged**